### ISSUE TITLE
add the ability to pass "-cpu host" option in xen-hvm

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#215]
Originally reported of Google Code with ID 159.

```
There should be a way to pass "-cpu $cpuflags" option to kvm.

In my use case, having a 2 node cluster of identical hardware, passing the "-cpu host" option to KVM would be a neat performance optimisation.

( as per http://www.linux-kvm.org/page/Tuning_KVM )

Thinking out loud, not sure how realstic this is:

When using different hardware/cpu's on the nodes, it might be interesting to configure a common feature set supported on all nodes? Maybe this could be implementyed with some special "auto flag, and let ganeti determine this?

```
Originally added on 2011-05-12 12:05:02 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
In the mean time, one can use a wrapper script. To have kvm execute with the "-cpu host" flag, one can use following procedure, to be exectued on all nodes:

dpkg-divert --add --rename --divert /usr/bin/kvm.real /usr/bin/kvm
cat  <<EOF > /usr/bin/kvm
> #!/bin/sh
> 
> /usr/bin/kvm.real -cpu host “\$@”
> 
> EOF
chmod +x /usr/bin/kvm
```
Originally added on 2011-05-12 12:54:16 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
The way Ganeti is thought to be deployed, all machines in the cluster should be somewhat homogeneous. Different disk sizes or memory sizes are fine, but not different CPU capabilities, as that will break migration to a bigger or lesser degree anyway.

So yes, -cpu host makes sense, thanks for reporting.
```
Originally added on 2011-05-12 13:56:10 +0000 UTC.

Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
I'd also like to see this for xen. Xen can do limit the cpuid at the hypervisor level (boot options) or at the xen level (hostname.cfg).

http://zhigang.org/wiki/XenCPUID#cpuid-overridden-for-xen-guests

has further info.

This would allow nonhomogenous machines to form a cluster. The cluster verify command could check the xen hw_caps (xm info | grep hw_caps) of the lowest capable machine in the cluster and warn if any VMs are running with "illegal" capabilities (if they were to be migrated).

Likewise, migrate could block if the instances capabilities are not a subset of the target nodes capabilities. Failover would be fine, as it restarts the instance. As a backend parameter, it could be set on a per-instance basis, with the cluster default reducing any time a node with lower capabilities is added.

As a starting point, adding a "generic i686 class" by default like in the above link, would allow for much better migration of hosts on nonhomogenous machines.
```
Originally added on 2012-01-19 05:48:56 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
In case anyone else is looking for this for xen, we are currently using the attached patch to make VMs appear as generic i686 cpu.

Before:
# cat /proc/cpuinfo 
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 23
model name      : Intel(R) Xeon(R) CPU           X3350  @ 2.66GHz
stepping        : 7
cpu MHz         : 2660.054
cache size      : 6144 KB
fpu             : yes
fpu_exception   : yes
cpuid level     : 10
wp              : yes
flags           : fpu de tsc msr pae cx8 sep cmov pat clflush mmx fxsr sse sse2 ss ht syscall nx lm constant_tsc up rep_good aperfmperf pni ssse3 cx16 sse4_1 hypervisor lahf_lm
bogomips        : 5320.10
clflush size    : 64
cache_alignment : 64
address sizes   : 36 bits physical, 48 bits virtual
power management:

After:
# cat /proc/cpuinfo 
processor       : 0
vendor_id       : unknown
cpu family      : 6
model           : 11
model name      : 06/0b
stepping        : 1
cpu MHz         : 2660.054
fpu             : yes
fpu_exception   : yes
cpuid level     : 3
wp              : yes
flags           : fpu de tsc msr cx8 sep cmov clflush mmx up hypervisor
bogomips        : 5320.10
clflush size    : 64
cache_alignment : 64
address sizes   : 36 bits physical, 48 bits virtual
power management:

It would be nice if ganeti could figure out the "lowest" cpuid needed to allow all VMs to live migrate successfully, but for now this seems to work for us.
```
Originally added on 2012-02-13 07:09:46 +0000 UTC.

Attachments: 
[ganeti-xen-i686.diff](https://storage.googleapis.com/google-code-attachments/ganeti/issue-159/comment-4/ganeti-xen-i686.diff)

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2012-07-19 13:17:09 +0000 UTC.

Added Labels: Type-Enhancement Hypervisor-KVM

**Comment by Ganeti-Issues-Migrator** [comment]
```
This was implemented for KVM, and will be released in 2.7. We still need to do the Xen-specific parts, though.
```
Originally added on 2012-08-11 11:53:56 +0000 UTC.

Added to Milestone: Release2.7

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-03-01 19:32:19 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Retargeting at 2.8
```
Originally added on 2013-04-08 14:06:20 +0000 UTC.

Added to Milestone: Release2.8
Removed from Milestone: Release2.7

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-05-08 08:25:55 +0000 UTC.

Added Labels: SmallTask

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-08-20 08:26:28 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-08-20 08:29:19 +0000 UTC.

Added to Milestone: Release2.9
Removed from Milestone: Release2.8

**Comment by Ganeti-Issues-Migrator** [comment]
```
CPU support has been written (commit 295ed29) and should be released on Ganeti 2.10
```
Originally added on 2013-10-19 08:51:49 +0000 UTC.

Changed State: Fixed
Added to Milestone: Release2.10
Removed from Milestone: Release2.9

**Comment by Ganeti-Issues-Migrator** [comment]
```
2.10.0 with fixes for these bugs has just been released.
```
Originally added on 2014-02-20 13:25:23 +0000 UTC.

Changed State: Released
